mm2src/kdf_walletconnect/src/lib.rs

```diff
@@ -0,0 +1,564 @@
+            }
+        }
+
+        MmError::err(WalletConnectError::InternalError(
+            "[{topic}] client connection timeout".to_string(),
+        ))
+    }
+
+    /// This persistent reconnection and retry strategy keeps the WebSocket connection active,
+    /// allowing the client to automatically resume operations after network interruptions or disconnections.
+    /// Since TCP handles connection timeouts (which can be lengthy), we're using a shorter timeout here
+    /// to detect issues quickly and reconnect as needed.
+    async fn wait_until_client_is_online_loop(&self, attempt: usize) {
+        debug!("Attempt {} failed due to timeout. Reconnecting...", attempt + 1);
+        loop {
+            match self.reconnect_and_subscribe().await {
+                Ok(_) => {
+                    info!("Reconnected and subscribed successfully.");
+                    break;
+                },
+                Err(reconnect_err) => {
+                    error!("Reconnection attempt failed: {reconnect_err:?}. Retrying...");
+                    Timer::sleep(1.5).await;
+                },
+            }
+        }
+    }
+
+    /// Checks if the current session is connected to a Ledger device.
+    /// NOTE: for COSMOS chains only.
+    pub fn is_ledger_connection(&self) -> bool {
+        self.session_manager
+            .get_session_active()
+            .and_then(|session| session.session_properties)
+            .and_then(|props| props.keys.as_ref().cloned())
+            .and_then(|keys| keys.first().cloned())
+            .map(|key| key.is_nano_ledger)
+            .unwrap_or(false)
+    }
+
+    /// Checks if a given chain ID is supported.
+    pub(crate) fn validate_chain_id(
+        &self,
+        session: &Session,
+        chain_id: &WcChainId,
+    ) -> MmResult<(), WalletConnectError> {
+        if let Some(Namespace {
+            chains: Some(chains), ..
+        }) = session.namespaces.get(chain_id.chain.as_ref())
+        {
+            if chains.contains(&chain_id.to_string()) {
+                return Ok(());
+            };
+        }
+
+        // https://specs.walletconnect.com/2.0/specs/clients/sign/namespaces#13-chains-might-be-omitted-if-the-caip-2-is-defined-in-the-index
+        if session.namespaces.contains_key(&chain_id.to_string()) {
+            return Ok(());
+        }
+
+        MmError::err(WalletConnectError::ChainIdNotSupported(chain_id.to_string()))
+    }
+
+    pub async fn validate_update_active_chain_id(&self, chain_id: &WcChainId) -> MmResult<(), WalletConnectError> {
+        let session =
+            self.session_manager
+                .get_session_active()
+                .ok_or(MmError::new(WalletConnectError::SessionError(
+                    "No active WalletConnect session found".to_string(),
+                )))?;
+
+        self.validate_chain_id(&session, chain_id)?;
+
+        // TODO: uncomment when WalletConnect wallets start listening to chainChanged event
```

Comment:
What is needed to be implemented to support listening to `chainChanged` events? Ideally it should be in this PR.

Reply:
Metamask will automatically prompt user to update their active chain/account if it differs from the chain/account used to activate a coin in KDF during chain/account specific request like transaction signing request etc.  Same as trust wallet I believe.

So nothing happens(I mean wallets don't listen to this event) if we send such request

Reply:
> So nothing happens(I mean wallets don't listen to this event) if we send such request

so what about just sending it for now and when wallets listen to it later we get this feat for free